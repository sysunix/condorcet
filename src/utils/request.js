import firebase, { db } from "../firebase";

export const createPoll = async poll => {
  const pollWithTimestap = {
    ...poll,
    timestamp: firebase.firestore.FieldValue.serverTimestamp()
  };

  return db.collection("polls").add(pollWithTimestap);
};

export const createVote = async (vote, pollId, userId) => {
  const voteWithTimestap = {
    ...vote,
    timestamp: firebase.firestore.FieldValue.serverTimestamp()
  };

  return db
    .collection("polls")
    .doc(pollId)
    .collection("votes")
    .doc(userId)
    .set(voteWithTimestap);
};
