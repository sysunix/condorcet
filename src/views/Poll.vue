<template>
  <div>
    <p class="text-center">{{ $t("poll.dragAndDrop") }}</p>

    <SortableList
      v-model="answers"
      class="mt-6 mx-auto w-full px-2"
      lock-axis="y"
      @input="onMove"
    >
      <SortableItem
        v-for="(answer, index) in answers"
        :key="answer.rank"
        :index="index"
        :rank="answer.rank"
        :value="answer.value"
      ></SortableItem>
    </SortableList>

    <button
      data-testid="vote"
      class="mt-4 w-full bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded"
      @click="vote"
    >
      {{ $t("global.vote") }}
    </button>
  </div>
</template>

<script>
import { mapState, mapActions, mapGetters } from "vuex";
import shuffle from "shuffle-array";
import SortableList from "../components/SortableList";
import SortableItem from "../components/SortableItem";
import { createVote } from "../utils/request";

export default {
  name: "Poll",
  components: {
    SortableList,
    SortableItem
  },
  data() {
    return {
      answers: []
    };
  },
  computed: {
    ...mapGetters("user", ["userId"]),
    ...mapState({
      poll: state => state.poll.current,
      userId: state => state.user.id
    })
  },
  async mounted() {
    await this.setPoll(this.$route.params.id);

    if (!this.poll.users.includes(this.userId)) {
      this.$router.push({ name: "polls_list" });
    }

    this.answers = shuffle(this.poll.answers).map((answer, index) => {
      return {
        rank: index + 1,
        ...answer
      };
    });
  },
  methods: {
    ...mapActions("poll", ["setPoll", "fetchPoll"]),
    ...mapActions("app", ["addNotification"]),
    onMove(inputs) {
      this.answers = inputs.map((input, index) => {
        return {
          ...input,
          rank: index + 1
        };
      });
    },
    async vote() {
      const pollId = this.$route.params.id;
      const userId = this.userId;

      try {
        await createVote({ vote: this.answers }, pollId, userId);

        this.addNotification({
          text: "Merci de ton vote",
          status: "success"
        });

        this.$router.push({
          name: "polls_list"
        });
      } catch (error) {
        this.addNotification({
          text: "Il y a eu un probl√®me",
          status: "error"
        });
      }
    }
  }
};
</script>

<style></style>
