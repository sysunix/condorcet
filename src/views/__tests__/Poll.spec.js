import { shallowMount, createLocalVue } from "@vue/test-utils";
import Vuex from "vuex";
import shuffle from "shuffle-array";
import { createVote } from "../../utils/request";

import Poll from "../Poll.vue";
import {
  MOCK_POLL_ID,
  MOCK_USER_ID,
  MOCK_ANSWERS,
  getTranslation
} from "../../utils/test";

jest.mock("shuffle-array");
jest.mock("../../firebase");
jest.mock("../../utils/request");

shuffle.mockReturnValue(MOCK_ANSWERS);

const $route = {
  params: { id: MOCK_POLL_ID }
};

const localVue = createLocalVue();
localVue.use(Vuex);

describe("Poll.vue", () => {
  let wrapper;
  let appActions;
  let pollActions;

  beforeEach(() => {
    jest.clearAllMocks();
    appActions = { addNotification: jest.fn() };
    pollActions = { setPoll: jest.fn() };

    const store = new Vuex.Store({
      state: {
        user: { id: MOCK_USER_ID }
      },
      modules: {
        poll: {
          namespaced: true,
          state: {
            current: { answers: MOCK_ANSWERS, users: [MOCK_USER_ID] }
          },
          actions: pollActions
        },
        app: { namespaced: true, actions: appActions }
      }
    });

    wrapper = shallowMount(Poll, {
      localVue,
      store,
      mocks: {
        $t: getTranslation,
        $route,
        $router: { push: jest.fn() }
      }
    });
  });

  it("should render", () => {
    expect(pollActions.setPoll).toBeCalled();
    expect(wrapper).toMatchSnapshot();
  });

  it("click to inform user with a notification", () => {
    const voteButton = wrapper.find("button");

    voteButton.trigger("click");

    expect(createVote).toBeCalled();
  });
});
