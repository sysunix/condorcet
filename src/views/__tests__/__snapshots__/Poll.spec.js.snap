// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Poll.vue should render 1`] = `
<div>
  <div class="text-center text-4xl"></div>
  <p class="text-center mt-3">Pour voter il suffit de classer vos préférences en les glissant-déposant</p>
  <sortablelist-stub value="[object Object],[object Object],[object Object]" axis="y" distance="0" pressdelay="0" pressthreshold="5" hidesortableghost="true" lockoffset="50%" transitionduration="300" appendto="body" lockaxis="y" shouldcancelstart="function _default(e) {
        // Cancel sorting if the event target is an \`input\`, \`textarea\`, \`select\` or \`option\`
        var disabledElements = ['input', 'textarea', 'select', 'option', 'button'];
        return disabledElements.indexOf(e.target.tagName.toLowerCase()) !== -1;
      }" gethelperdimensions="function _default(_ref) {
        var node = _ref.node;
        return {
          width: node.offsetWidth,
          height: node.offsetHeight
        };
      }" class="mt-6 mx-auto w-full px-2">
    <sortableitem-stub index="0" collection="default" value="Luffy" rank="1"></sortableitem-stub>
    <sortableitem-stub index="1" collection="default" value="Zorro" rank="2"></sortableitem-stub>
    <sortableitem-stub index="2" collection="default" value="Sanji" rank="3"></sortableitem-stub>
  </sortablelist-stub> <button data-testid="vote" class="mt-4 w-full bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded">
    Voter
  </button>
</div>
`;
